File: python/recursions_backtrackings/day3/recursion.py
# # reverse an array...
# def reverse(l, r, arr):
#     if (l >= r): 
#         return
#     arr[l], arr[r] = arr[r], arr[l]

#     reverse(l + 1, r - 1, arr)


# # -------- main --------
# n = int(input("Enter size of array: "))

# arr = list(map(int, input("Enter array elements: ").split()))

# reverse(0, n - 1, arr)

# print("Reversed array:", arr)






# # reverse an array...
# def reverse(i, arr):
#     n = len(arr)
#     if i >= n // 2:
#         return
#     arr[i], arr[n - i - 1] = arr[n - i - 1], arr[i]

#     reverse(i + 1, arr)


# # -------- main --------
# n = int(input("Enter size of array: "))

# arr = list(map(int, input("Enter array elements: ").split()))

# reverse(0, arr)

# print("Reversed array:", arr)






# check if a string is palindrome or not...
def palindrome(i, s):
    n = len(s)
    if i >= n // 2:
        return True
    
    if s[i] != s[n - i - 1]:
        return False
    
    return palindrome(i + 1, s)


# -------- main --------
s = input("Enter a string: ")

if palindrome(0, s):
    print("String is a Palindrome")
else:
    print("String is NOT a Palindrome")

File: java/recursions_backtrackings/Day3/Recursion.java
package recursions_backtrackings.Day3;
import java.util.*;

public class Recursion {

    // Reverse an array...
    public static void reverse(int l, int r, int[] arr) {
        if (l >= r) return;

        int temp = arr[l];
        arr[l] = arr[r];
        arr[r] = temp;

        reverse(l + 1, r - 1, arr);
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter the size of array: ");
        int n = sc.nextInt();

        int[] arr = new int[n];

        System.out.print("Enter array elements: ");
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }

        reverse(0, n - 1, arr);

        System.out.print("Reversed array: ");
        for (int i = 0; i < n; i++) {
            System.out.print(arr[i] + " ");
        }

        sc.close();
    }

}
